src/Client.ts

```diff
@@ -344,53 +356,88 @@ export default class Client extends EventEmitter {
                 this.sm,
                 this.stanzas
             );
-            this.transport.connect({
-                acceptLanguages: this.config.acceptLanguages || ['en'],
+
+            const config: TransportConfig = {
+                acceptLanguages: this.config.acceptLanguages || [this.config.lang ?? 'en'],
                 jid: this.config.jid!,
-                lang: this.config.lang || 'en',
+                lang: this.config.lang ?? 'en',
                 server: this.config.server!,
-                url: conf.url!,
-                ...conf
-            });
-            return;
+            };
+
+            if (typeof settings === 'string') {
+                transportEndpoints.push([transport, { ...config, url: settings }]);
+            } else if (settings === true) {
+                if (transport.discoverBindings) {
+                    for (const ep of (await transport.discoverBindings(this.config.server!) ?? [])) {
+                        transportEndpoints.push([transport, { ...config, ...ep }]);
+                    }
+                } else {
+                    for (const ep of (endpoints[name] ?? [])) {
+                        transportEndpoints.push([transport, { ...config, url: ep }]);
+                    }
+                }
+            } else if (typeof settings === 'object') {
+                transportEndpoints.push([transport, { ...config, ...settings }]);
+            }
+        }
+
+        const secureOptions: Array<[Transport, TransportConfig]> = [];
+        const insecureOptions: Array<[Transport, TransportConfig]> = [];
+
+        // secureOptions + insecureOptions will be sorted as transportEndpoints
+        // is created in priority order.
+        for (const [transport, endpoint] of transportEndpoints) {
+            if (
+                endpoint.url?.startsWith('https://') ||
+                endpoint.url?.startsWith('wss://') ||
+                transport instanceof TCP
+            ) {
+                secureOptions.push([transport, endpoint]);
+            } else {
+                insecureOptions.push([transport, endpoint]);
+            }
+        }
+
+        const options = 
+            this.config.requireSecureTransport
+            ? secureOptions
+            : secureOptions.concat(insecureOptions);
+
+        for (const [transport, endpoint] of options) {
+            this.transport = transport;
+            const dcPromise = new Promise(resolve => this.once('disconnected', resolve));
+            try {
+                await this.transport.connect(endpoint);
+                return;
+            } catch (_) {
+                await this.disconnect();
+                this.transport = undefined;
+            }
+            await dcPromise;
         }
 
-        console.error('No endpoints found for the requested transports.');
         this.emit('--transport-disconnected');
+        throw 'No adequate endpoints found for the requested transports.';
```

Comment:
Wouldn't this throw if the user purposefully disconnects, too?

Reply:
No, that's just the fall through case for no connection being made successfully. Maybe disconnecting somehow _during_ connection will mess up the current transport and cause this, but in normal usage this only fires when the early return isn't hit, and only during connection.